KEP 3140: TimeZone support in CronJob (#3143)
kubernetes/enhancements: keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md

Zone used during the creation depens on where kube-controller-manager is running.
This proposal aims to extend CronJob resource with the ability for a user to
define the TimeZone when a Job should be created.

## Motivation

Not long after the [introduction of CronJob in kubernetes](https://github.com/kubernetes/kubernetes/pull/11980)
a [request was raised to support setting time zones](https://github.com/kubernetes/kubernetes/issues/47202).
The initial [response from SIG-Apps and SIG-Architecture](https://github.com/kubernetes/kubernetes/issues/47202#issuecomment-360820586)
at the time was that introducing such functionality would require cluster operator
to manually include TimeZone database since golang did not have one.
Starting from [golang 1.15](https://go.dev/doc/go1.15) `time/tzdata` package can
be embedded in the binary thus removing the requirement for external database.
At that time the majority of the focus was towards [moving CronJob to GA](https://github.com/kubernetes/enhancements/issues/19)
so the effort to support TimeZone was again delayed. Now that we have CronJob
fully GA-ed it's time to satisfy the original request.

### Goals

> - Add the field `.spec.timeZone` which allows specifying a valid TimeZone

Reviewer: Is this a name or an offset from UTC?

Author: name - updated

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

Add the field `.spec.timeZone` to the CronJob resource. The cronjob controller
will take the field into account when scheduling the next Job run. In case the

> field is not specified or is empty the controller will maitain the current

Reviewer 2: ```suggestion
field is not specified or is empty the controller will maintain the current
```

behavior which is to rely on the time zone of the kube-controller-manager
process.

### Notes/Constraints/Caveats (Optional)

The current mechanism, which will still be the default behavior heavily relies
on the time zone of the kube-controller-manager process which is hard for a
regular user to figure out. Exposing an explicit field for setting a time zone
will allow CronJob authors to simplify the user experience when creating CronJobs.

> 

Reviewer 3: Would it make sense to also introduce a way to set a global default TZ?

Reviewer: Ah, I had an idea about that.
_In my view_ that is best left for an external controller or for an admission webhook. However, the “external” mechanism could still be official Kubernetes code.

Author: Yes, no global setting, we don't have it anywhere, and I'd leave it to cluster admin to implement, re-use such setting.

### Risks and Mitigations

- Outdated time zone in the golang might lead to wrong schedule times

This problem can be mitigated with a fresh build of kube-controller-manager with
updated golang version, but it heavily relies on the go community to keep the
time zone database up-to-date.

- Malicious user can create multiple CronJobs with different time zone which can

> actually trigger Jobs at the exact same time

Reviewer 2: Can't malicious user trigger Jobs at the same time without time zone?

Author: It's easier to notice if all jobs are expressed w/o timezone, with timezone it's slightly harder.

Cluster administrators should enforce quota to ensure not that many Jobs and
CronJobs can be created per user.

## Design Details

### CronJob API

> The `CronJobSpec` structure is expanded with new `TimeZone` field which allows
> specifying the name of the time zone to be used. Missing or empty value of the
> field indicates the current behavior, which relies on the time zone of the
> kube-controller-manager process.

Reviewer: ```suggestion
The `.spec` for a CronJob is expanded with new `timeZone` field which allows
specifying the name of the time zone to be used. Missing or empty value of the
field indicates the current behavior, which relies on the time zone of the
kube-controller-manager process.

In the API code, that looks like:
```

1. Does the kube-controller-manager populate the `.spec.timeZone` once it detects a CronJob with no timezone set?
2. Should it?

```golang

> type CronJobSpec struct {
> 
> 	// The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
> 	Schedule string
> 
>   // Time zone for the above schedule
>   TimeZone *string
> 
> }

Reviewer 2: ```suggestion
type CronJobSpec struct {

	// The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
	Schedule string

	// Time zone for the above schedule
	TimeZone *string

}
```

Author: Fixed.

```

The value provided in `TimeZone` field will be validated against the embedded

> golang timezone database.

Reviewer: Does this mean that changes to time zone databases need to be applied to the API server and kube-controller-manager in a particular order?

Author: No need for manual action, just follow the general guideline for updating the cluster: https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/

    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.

> -->

Reviewer: Whichever sense of the `.spec.timeZone` field we use (name or offset from UTC), let's list the other approach as an alternative.

Author: Done.

## Motivation

Not long after the [introduction of CronJob in kubernetes](https://github.com/kubernetes/kubernetes/pull/11980)
a [request was raised to support setting time zones](https://github.com/kubernetes/kubernetes/issues/47202).
The initial [response from SIG-Apps and SIG-Architecture](https://github.com/kubernetes/kubernetes/issues/47202#issuecomment-360820586)
at the time was that introducing such functionality would require cluster operator
to manually include TimeZone database since golang did not have one.
Starting from [golang 1.15](https://go.dev/doc/go1.15) `time/tzdata` package can
be embedded in the binary thus removing the requirement for external database.
At that time the majority of the focus was towards [moving CronJob to GA](https://github.com/kubernetes/enhancements/issues/19)
so the effort to support TimeZone was again delayed. Now that we have CronJob
fully GA-ed it's time to satisfy the original request.

### Goals

- Add the field `.spec.timeZone` which allows specifying a valid TimeZone name

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

> Add the field `.spec.timeZone` to the CronJob resource. The cronjob controller

Reviewer: We could make the field a bit more self-documenting
```suggestion
Add the field `.spec.timeZoneName` to the CronJob resource. The cronjob controller
```

Author: I think using time zone is in most cases sufficient.

updated golang version, but it heavily relies on the go community to keep the
time zone database up-to-date.

- Malicious user can create multiple CronJobs with different time zone which can
actually trigger Jobs at the exact same time

Cluster administrators should enforce quota to ensure not that many Jobs and
CronJobs can be created per user.

## Design Details

### CronJob API

The `.spec` for a CronJob is expanded with new `timeZone` field which allows
specifying the name of the time zone to be used. Missing or empty value of the
field indicates the current behavior, which relies on the time zone of the
kube-controller-manager process.

In the API code, that looks like:

```golang

type CronJobSpec struct {

    // The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
    Schedule string

    // Time zone for the above schedule
    TimeZone *string

}
```

The value provided in `TimeZone` field will be validated against the embedded
golang timezone database.

### CronJob controller

CronJob controller will use non-nil, non-empty value from `TimeZone` field when
parsing the schedule and during scheduling the next run time. Additionally the

> time zone will be reflected in the `.status.lastSuccessfulTime` and `.status.lastScheduleTime`.

Reviewer: We could also record the applied offset from UTC (from the last invocation) in `.status`. Maybe after an alpha though?

Author: `lastSuccessfulTime` is of `metav1.Time` type which embeds `time.Time` which already has time zone information embeded in it, so I'm planning to re-use that.

- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

- Upgrades

When upgrading from a release without this feature to a release with `TimeZone`
users should not notice any change in behavior. The default value for `TimeZone`
is nil, which is equal to current behavior for backwards compatibility.

- Downgrades

When downgrading from a release with this feature, to a release without `TimeZone`
there are a few cases:

>   1. If TimeZone feature gate was enabled and user specified a TimeZone, the newly

Reviewer 4: How will you test to be sure the cronjob controller correctly handles going from timezone aware to non-timezone aware?

Author: Manual testing of a cronjob with time zone specified, I can't think of a better approach.

Author: I've mentioned that in the test plan, that we want to enabled with the feature on and off to ensure the controller behaves as expected. As much as possible those tests will be automated in integration, but not everything is possible there.

  created Jobs after downgrade will return to previous behavior, as if no TimeZone
  was ever specified.
  2. If TimeZone feature gate was enabled and user did not specify TimeZone, there
  should be no change in behavior.
  3. If TimeZone feature gate was not enabled there should be no change in behavior.

### Version Skew Strategy

This feature has no node runtime implications.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: CronJobTimeZone
  - Components depending on the feature gate: kube-apiserver, kube-controller-manager

###### Does enabling the feature change any default behavior?

No, the default behavior is maintained independently of the feature gate.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes, the feature can be disabled. The outcome of disabling will depend if the

Reviewer 4: CronJob controller has had difficulties in the past.  How will you test the enable, disable, enable flow?

Author: Similarly as with the previous, a manual testing will be performed to ensure this works as expected.

Reviewer 4: unit tests from line 272 will work.

new field was set by the user:

1. If there was no value set in `TimeZone` field there will be no change in behavior.
2. If the user set a valid `TimeZone`, the newly created Jobs will be triggered
as if the field was never set.

###### What happens if we reenable the feature if it was previously rolled back?

The controller will start reading the new field.

###### Are there any tests for feature enablement/disablement?

Yes, both units and integration tests for enablement, disablement and transitions.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

> ###### How can a rollout or rollback fail? Can it impact already running workloads?

Reviewer 4: An upgrade flow can be vulnerable to the enable, disable, enable if you have a lease that is acquired by a new KCM, then an old KCM, then a new KCM.  The unit and integration tests need to be robust.

Author: ack, added that here

and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

- [ ] Events
  - Event Reason:
- [ ] API .status
  - Condition name:
  - Other field:
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

> - [ ] Metrics

Reviewer 4: While not required for alpha, can you think of what metrics you'll watch?  If you're able to fill this in now, people running the alpha will know what information to look for.

      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> ## Alternatives

Reviewer: I suppose the other other alternative is timezone details in the cron time spec (the approach that got merged by accident).

Author: Rarely, we re-use the current fields, that case you're talking about is more a feature creep not intentional on our side, b/c it exposes implementation details, that were not covered in the original design .

Reviewer: The inline timezone spec - as mentioned in https://web.archive.org/web/20211007074332/https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/ - is not the design we want. I get that. We should nonetheless list it as a ruled-out alternative.

Reviewer: I can PR that in after this merges.

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

CronJob creates Jobs based on the schedule specified by the author, but the Time
Zone used during the creation depends on where kube-controller-manager is running.
This proposal aims to extend CronJob resource with the ability for a user to
define the TimeZone when a Job should be created.

## Motivation

Not long after the [introduction of CronJob in kubernetes](https://github.com/kubernetes/kubernetes/pull/11980)
a [request was raised to support setting time zones](https://github.com/kubernetes/kubernetes/issues/47202).
The initial [response from SIG-Apps and SIG-Architecture](https://github.com/kubernetes/kubernetes/issues/47202#issuecomment-360820586)
at the time was that introducing such functionality would require cluster operator
to manually include TimeZone database since golang did not have one.
Starting from [golang 1.15](https://go.dev/doc/go1.15) `time/tzdata` package can

> be embedded in the binary thus removing the requirement for external database.

Reviewer 5: I think we should explicitly call out in the KEP intent to embed this package as the approach for keeping tzdata up to date?

(Or else, if we have another approach planned, we should outline it?)

Reviewer: Hmm. If there is a timezone change (eg a new national law), does that trigger a patch release?

Reviewer: The approach I initially put forward was to have Kubernetes itself only aware of offsets from UTC.

Reviewer 5: The implementing PR appears to be taking the `import _ "time/tzdata"` approach, operators would have the option to patch their deployments with images or mounts containing a non-embedded on-disk tzdata and go will prefer that.

I'm also not sure that the embedded copy will be used at all in official releases / it's not clear how helpful importing this is, all of our images should be shipping with tzdata on-disk in the image. will x-ref the PR.

Author: I've already responded to that other comment, but I'll leave the justification also here. The intent why we picked this approach, as a safe net was mentioned here in the motivation section. I'll make sure to add some more details during beta reviews to provide more details around expectations and assumptions for the embedded vs the on-system tz database.
